use clean architecture approuch

olegdaniv opened the pull request:
- fixed marge conflict
- added UseCases, refactored requests, used clean architecture

Commit:
- fixed Failed approach
- fixed marge conflict
- added UseCases, refactored requests, used clean architecture

iDobro commented on app/src/main/java/com/example/topmovies/datalayer/network/MoviesApi.kt:
```diff
@@ -0,0 +1,53 @@
+    @GET("https://86e6737b-9ab8-444a-9c53-bdfa86309d8f.mock.pstmn.io/movies")
+    fun getMovies(): Call<MovieObject>
+
+    data class MovieObject(
```
Move this to own file, don't put models into API interfaces.

iDobro commented on app/src/main/java/com/example/topmovies/datalayer/network/MoviesApi.kt:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    data class MovieApi(
```
Bad naming

iDobro commented on app/src/main/java/com/example/topmovies/datalayer/dao/MoviesDao.kt:
```diff
@@ -0,0 +1,67 @@
+        val imDbRatingCount: String,
+        var isFavorite: Boolean = false
+    ) {
+        fun toMovie(): Movie {
```
Need to find a better way to wrap data models from the data layer to the presentation or domain.

olegdaniv replied:
I'm going to do this in next PR

iDobro commented on app/src/main/java/com/example/topmovies/datalayer/repository/MovieDetailsRepository.kt:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    data class MovieDetails(
```
Move to own file

iDobro commented on app/src/main/java/com/example/topmovies/datalayer/repository/MovieDetailsRepository.kt:
```diff
@@ -0,0 +1,75 @@
+
+    fun getMovieDetails(id: String): ResultOf<Failure, MovieDetails>
+
+    fun insertMovieDetails(
```
Something is wrong, this method isn't used.

iDobro commented on app/src/main/java/com/example/topmovies/di/Modules.kt:
```diff
@@ -43,19 +48,10 @@ val networkModule = module {
-    single { UpdateFavoriteMovieUseCase(get(), get(), get()) }
-    single { LoadMoviesUseCase(get(), get(), get()) }
-    single { GetMovieDetailsUseCase(get(), get(), get()) }
+    single(named("moviesR")) { get<Retrofit>().create(MoviesApi::class.java) }
```
Rename "R"

iDobro commented on app/src/main/java/com/example/topmovies/ui/fragment/MovieDetailsFragment.kt:
```diff
@@ -0,0 +1,64 @@
+class MovieDetailsFragment : BaseFragment() {
+
+    companion object {
+        const val FRAGMENT_KEY = "movieID"
```
* ID_ARGUMENT_KEY="id_argument_key"

iDobro commented on app/src/main/java/com/example/topmovies/ui/viewmodel/BaseViewModel.kt:
```diff
@@ -0,0 +1,16 @@
+
+abstract class BaseViewModel : ViewModel() {
+
+    private val _errorMessage = SingleLiveEvent<Failure>()
```
* SingleEventLiveData

iDobro commented on app/src/main/java/com/example/topmovies/ui/viewmodel/MovieViewModel.kt:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    private fun foldNewResult(
```
Try to find better solution

iDobro commented on app/src/main/java/com/example/topmovies/utils/AuxiliaryFunction.kt:
```diff
@@ -0,0 +1,42 @@
+package com.example.topmovies.utils
+
```
Separate this file depending on usage.

iDobro commented on app/src/main/java/com/example/topmovies/utils/SingleLiveEvent.kt:
```diff
@@ -0,0 +1,34 @@
+import androidx.lifecycle.Observer
+import java.util.concurrent.atomic.AtomicBoolean
+
+class SingleLiveEvent<T> : MutableLiveData<T>() {
```
Need to clean up

Commit:
- Distributed files into packages

iDobro commented on app/src/main/java/com/example/topmovies/data/dao/MoviesDao.kt:
```diff
@@ -0,0 +1,47 @@
+    abstract fun updateMovieEntity(id: String, rank: String, rankUpDown: String)
+
+    @Transaction
+    open fun upsertMoviesEntity(movies: ResultOf<Failure, List<Movie>>) {
```
You don't need to use ResultOf here.

YuriiTsap replied:
I would agree with that.

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesRequest.kt:
```diff
@@ -0,0 +1,30 @@
+    private val api: MoviesApi,
+    private val getPreference: GetPreferenceUseCase,
+    private val networkHandler: NetworkHandler
+) : BaseRequest() {
```
Need to add empty line

Commit:
- took into account new comments

YuriiTsap commented on app/src/main/java/com/example/topmovies/data/dao/MoviesDao.kt:
```diff
@@ -0,0 +1,47 @@
+    @Query("SELECT * From tb_movies")
+    abstract fun getMoviesEntity(): List<MovieEntity>
+
+    @Query("Select * from tb_movies where id = :id")
```
why would you need to use the `tb_` prefix?

YuriiTsap commented on app/src/main/java/com/example/topmovies/data/dao/MoviesDao.kt:
```diff
@@ -0,0 +1,47 @@
+import com.example.topmovies.presentation.models.Movie
+
+@Dao
+abstract class MoviesDao {
```
Is there any particular reason for appending "Entity" to each method?
Looks a bit redundant to me.

YuriiTsap commented on app/src/main/java/com/example/topmovies/presentation/fragment/MovieDetailsFragment.kt:
```diff
@@ -0,0 +1,64 @@
+
+    private fun setupUI() {
+        viewModel.errorMessage.observe(viewLifecycleOwner) {
+            handlerFailure(it)
```
+

YuriiTsap commented on app/src/main/java/com/example/topmovies/presentation/fragment/MovieDetailsFragment.kt:
```diff
@@ -0,0 +1,64 @@
+        }
+        viewModel.movieDetails.observe(
+            viewLifecycleOwner,
+            ::renderMovieDetails
```
+

Commit:
- rename tableName without prefix "tb_"

Commit:
- renamed functions in DAO classes without "entity"

Commit:
- renamed generic in BaseRequest.kt and ResultOf.kt

Commit:
- removed swipeRefresh.isRefreshing from setupUi to setupViewmodel in MoviesFragment.kt

Commit:
- renamed generics in UseCase.kt

Commit:
- renamed missed function

Commit:
- import changed

Commit:
- changed abused "when"

Commit:
- Remove None.kt

Commit:
- renamed function run() to execute()
- created GetPairMoviesUseCase.kt

Commit:
- fixed GetPairMoviesUseCase.kt
- renamed ResultOf.kt to Result.kt
- deleted unused useCase and EnumResult.kt
- renamed some fields and functions

iDobro commented on app/src/main/java/com/example/topmovies/data/network/BaseRequest.kt:
```diff
@@ -0,0 +1,27 @@
+    fun <Type, Result> request(
+        call: Call<Type>,
+        transform: (Type) -> Result,
+    ): com.example.topmovies.domain.utils.Result<Failure, Result> {
```
Use static import

iDobro commented on app/src/main/java/com/example/topmovies/data/network/BaseRequest.kt:
```diff
@@ -0,0 +1,27 @@
+import okio.IOException
+import retrofit2.Call
+
+abstract class BaseRequest {
```
Bad naming, this class is not "Request", this class provides a method to transform (or unwrap) API response to your model. Need to find a better name.

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MovieDetailsAPi.kt:
```diff
@@ -0,0 +1,21 @@
+interface MovieDetailsAPi {
+
+    companion object {
+        const val MOVIE_DETAILED = "/en/API/Title"
```
In my opinion, all "companion objects" should be at the end of the file.

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MovieDetailsRequest.kt:
```diff
@@ -0,0 +1,27 @@
+import com.example.topmovies.presentation.models.MovieDetails
+
+class MovieDetailsRequest(
+    private val api: MovieDetailsAPi,
```
Typo in name :
MovieDetailsAPi -> MovieDetailsApi

iDobro commented on app/src/main/java/com/example/topmovies/data/network/BaseRequest.kt:
```diff
@@ -0,0 +1,27 @@
+
+abstract class BaseRequest {
+
+    fun <Type, Result> request(
```
Also to provide this method, we can use Extension Function from "Call".

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesApi.kt:
```diff
@@ -0,0 +1,21 @@
+
+interface MoviesApi {
+
+    companion object {
```
To end of file

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MovieDetailsRequest.kt:
```diff
@@ -0,0 +1,27 @@
+
+class MovieDetailsRequest(
+    private val api: MovieDetailsAPi,
+    private val getPreference: GetPreferenceUseCase,
```
Bad naming:
It's not UseCase.  

Unused reference: If you don't use it: 
1) Remove it or 2) Live comment, why it is here.

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesRequest.kt:
```diff
@@ -0,0 +1,32 @@
+
+class MoviesRequest(
+    private val api: MoviesApi,
+    private val getPreference: GetPreferenceUseCase,
```
*repeated mistake

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesRequest.kt:
```diff
@@ -0,0 +1,32 @@
+    private val networkHandler: NetworkHandler
+) : BaseRequest() {
+
+    companion object {
```
To end of file

iDobro commented on app/src/main/java/com/example/topmovies/data/repository/MovieDetailsRepository.kt:
```diff
@@ -0,0 +1,36 @@
+
+    fun loadNewMovieDetailsById(id: String): Result<Failure, MovieDetails>
+
+    class MovieDetailsRepositoryImp(
```
Typo in the name:
*impl

iDobro commented on app/src/main/java/com/example/topmovies/data/utils/NetworkHandler.kt:
```diff
@@ -1,31 +1,23 @@
                 connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
             (capabilities != null && capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET))
         } else {
+            @Suppress("DEPRECATION")
```
Leave a comment when your min android SDK version will be increased, this code should be removed, or remove annotation.

iDobro commented on app/src/main/java/com/example/topmovies/di/Modules.kt:
```diff
@@ -4,27 +4,32 @@ import android.os.Handler
 import java.util.concurrent.Executors
 import java.util.concurrent.TimeUnit
 
 const val DATABASE_NAME: String = "movie_database"
```
For constants that are used only in file where they are declared use the "private" modification.

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesApi.kt:
```diff
@@ -0,0 +1,21 @@
+interface MoviesApi {
+
+    companion object {
+        const val TOP_100_MOVIES = "/en/API/MostPopularMovies"
```
For constants that are used only in the file where they are declared use the "private" modification.

iDobro commented on app/src/main/java/com/example/topmovies/domain/usecase/GetPreferenceUseCase.kt:
```diff
@@ -0,0 +1,21 @@
+import android.content.SharedPreferences
+import com.example.topmovies.data.network.MoviesRequest
+
+/** when using api.getMovies() and api.getMovieDetails() without any parameters, it connects to a mock server.
```
Usually, sentences start with a capitalized letter.

iDobro commented on app/src/main/java/com/example/topmovies/domain/usecase/UseCase.kt:
```diff
@@ -0,0 +1,31 @@
+    abstract val executor: ExecutorService
+    abstract val handler: Handler
+
+    abstract fun execute(params: Params): com.example.topmovies.domain.utils.Result<Failure, Result>
```
Use static import

iDobro commented on app/src/main/java/com/example/topmovies/domain/usecase/UseCase.kt:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    fun <Type1, Type2, Result> let(
```
Bad naming
Don't use numbers in names

iDobro commented on app/src/main/java/com/example/topmovies/domain/utils/Failure.kt:
```diff
@@ -0,0 +1,6 @@
+package com.example.topmovies.domain.utils
+
+sealed class Failure {
```
Bad naming, it's not failure, it's more look like an Error.

iDobro commented on app/src/main/java/com/example/topmovies/domain/utils/Result.kt:
```diff
@@ -0,0 +1,10 @@
+package com.example.topmovies.domain.utils
+
+sealed class Result<out Blunder, out Outcome> {
```
Rename generics.
If you want to use a word for naming but you not sure what it means, you can use this : 
https://context.reverso.net/translation/english-ukrainian/
Read examples of using a word for a better understanding of what that means.

iDobro commented on app/src/main/java/com/example/topmovies/presentation/adapter/MoviesAdapter.kt:
```diff
@@ -13,18 +13,22 @@ import com.bumptech.glide.request.target.CustomTarget
-import com.example.topmovies.utils.GlideApp
+import com.example.topmovies.presentation.models.Movie
+import com.example.topmovies.presentation.utils.GlideApp
+
```
Remove extra line

iDobro commented on app/src/main/java/com/example/topmovies/presentation/adapter/MoviesAdapter.kt:
```diff
@@ -13,18 +13,22 @@ import com.bumptech.glide.request.target.CustomTarget
     private val onFavoriteMovieClick: (String, Boolean) -> Unit
 ) : ListAdapter<Movie, MoviesAdapter.MovieViewHolder>(MovieDiffCallBack()) {
 
+    companion object {
```
1) To the end of file
2) private*

iDobro commented on app/src/main/java/com/example/topmovies/presentation/fragment/BaseFragment.kt:
```diff
@@ -0,0 +1,17 @@
+
+abstract class BaseFragment : Fragment() {
+
+    fun handlerFailure(failure: Failure) {
```
Bad naming
Function can't be named "handler".

iDobro commented on app/src/main/java/com/example/topmovies/presentation/fragment/MovieDetailsFragment.kt:
```diff
@@ -0,0 +1,64 @@
+
+class MovieDetailsFragment : BaseFragment() {
+
+    companion object {
```
To the end of file

iDobro commented on app/src/main/java/com/example/topmovies/presentation/fragment/MovieDetailsFragment.kt:
```diff
@@ -0,0 +1,64 @@
+        )
+    }
+
+    private fun renderMovieDetails(movieDetails: MovieDetails) = with(binding) {
```
**!!!
!!!Don't change any code that was accepted before and not included in the current task scope!!!
!!!**

iDobro commented on app/src/main/java/com/example/topmovies/presentation/fragment/SettingsFragment.kt:
```diff
@@ -8,11 +8,18 @@ import androidx.preference.ListPreference
 class SettingsFragment : PreferenceFragmentCompat() {
-    
+
+    companion object {
```
Repeated

iDobro commented on app/src/main/java/com/example/topmovies/presentation/utils/ThemeHandler.kt:
```diff
@@ -0,0 +1,23 @@
+
+class ThemeHandler {
+
+    companion object {
```
Repeated

iDobro commented on app/src/main/java/com/example/topmovies/presentation/viewmodel/BaseViewModel.kt:
```diff
@@ -0,0 +1,16 @@
+    private val _errorMessage = SingleEventLiveData<Failure>()
+    val errorMessage: LiveData<Failure> = _errorMessage
+
+    protected fun handledErrors(error: Failure) {
```
Bad naming

iDobro commented on app/src/main/java/com/example/topmovies/presentation/viewmodel/MovieViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+import com.example.topmovies.presentation.utils.EnumScreen.MOVIES
+
+class MovieViewModel constructor(
+    private val getPairMovies: GetPairMoviesUseCase,
```
*getMoviesPair
*GetMoviesPairUseCase

iDobro commented on app/src/main/java/com/example/topmovies/presentation/viewmodel/MovieViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+
+    fun getMovies() {
+        getPairMovies(Unit) {
+            handleResultOf(it)
```
Typo in word
Just "handleResult"

iDobro commented on app/src/main/java/com/example/topmovies/presentation/viewmodel/MovieViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    private fun shiftMovieToFavorites(id: String, isFavorite: Boolean) {
```
Inaccuracy:
shift -> switch or something like this

iDobro commented on app/src/main/java/com/example/topmovies/presentation/viewmodel/MovieViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    private fun handleResultOf(result: Result<Failure, Pair<List<Movie>, List<Movie>>>) {
```
Repeated bad naming

Commit:
- took the comments into PR

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MovieDetailsApi.kt:
```diff
@@ -0,0 +1,21 @@
+    fun getMovieDetails(): Call<MovieDetailsResponse>
+
+    companion object {
+        private const val MOVIE_DETAILED = "/en/API/Title"
```
Bad naming

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesApi.kt:
```diff
@@ -0,0 +1,21 @@
+    fun getMovies(): Call<MovieObjectResponse>
+
+    companion object {
+      private  const val TOP_100_MOVIES = "/en/API/MostPopularMovies"
```
Extra space

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MovieDetailsRequest.kt:
```diff
@@ -0,0 +1,29 @@
+
+class MovieDetailsRequest(
+    private val api: MovieDetailsApi,
+    private val getPreference: GetPreference,
```
Leave a comment, why is it here?

iDobro commented on app/src/main/java/com/example/topmovies/data/network/MoviesRequest.kt:
```diff
@@ -0,0 +1,27 @@
+
+class MoviesRequest(
+    private val api: MoviesApi,
+    /** Use this reference to obtain a key. **/
```
Please leave a comment, why we have this reference here but don't use it.

iDobro commented on app/src/main/java/com/example/topmovies/presentation/fragment/BaseFragment.kt:
```diff
@@ -0,0 +1,17 @@
+
+abstract class BaseFragment : Fragment() {
+
+    fun handleFailure(failure: Error) {
```
failure -> error

iDobro commented on app/src/main/java/com/example/topmovies/presentation/fragment/MovieDetailsFragment.kt:
```diff
@@ -0,0 +1,64 @@
+        )
+    }
+
+    private fun giveMovieDetails(movieDetails: MovieDetails) = with(binding) {
```
Maybe "showMovieDeteils"?

Commit:
- took the comments into PR

YuriiTsap commented on app/src/main/java/com/example/topmovies/data/repository/MovieDetailsRepository.kt:
```diff
@@ -0,0 +1,36 @@
+
+        override fun getMovieDetails(id: String): Result<Error, MovieDetails> {
+            val movieDetails = movieDetailsDao.getMovieDetailsById(id)
+            return if (movieDetails == null) {
```
Minor: I would prefer using let.

YuriiTsap commented on app/src/main/java/com/example/topmovies/domain/usecase/UseCase.kt:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    fun <TFirst, TSecond, Result> let(
```
Shouldn't be here.

YuriiTsap commented on app/src/main/java/com/example/topmovies/domain/usecase/UpdateFavoriteMovieUseCase.kt:
```diff
@@ -0,0 +1,34 @@
+        val pair = let(movies, favoriteMoves) { movies, favoriteMovies ->
+            Pair(movies, favoriteMovies)
+        }
+        return if (pair != null) {
+            Success(pair)
+        } else {
+            Failure(ServerError)
+        }
+    }
```
Use elvis for this kind of stuff.

YuriiTsap commented on app/src/main/java/com/example/topmovies/domain/utils/Result.kt:
```diff
@@ -0,0 +1,10 @@
+    data class Failure<out Error>(val error: Error) : Result<Error, Nothing>()
+    data class Success<out Data>(val result: Data) : Result<Nothing, Data>()
+
+    fun asSuccess() = this as Success<Data>
+
+    fun asErrors() = this as Failure<Error>
```
Should be as? 
Please double check the "fail scenario"

YuriiTsap commented on app/src/main/java/com/example/topmovies/domain/usecase/GetMoviesPairUseCase.kt:
```diff
@@ -0,0 +1,32 @@
+        val pair = let(movies, favoriteMoves) { movies, favoriteMovies ->
+            Pair(movies, favoriteMovies)
+        }
+        return if (pair != null) {
+            Success(pair)
+        } else {
+            Failure(ServerError)
+        }
+    }
+}
```
Please have a look at all related comments to if-else handling.

YuriiTsap commented on app/src/main/java/com/example/topmovies/presentation/fragment/MoviesFragment.kt:
```diff
@@ -75,32 +79,27 @@ class MoviesFragment : BaseFragment() {
+            MOVIES -> swipeRefresh.setOnRefreshListener {
+                moviesViewModel.loadNewMovies()
             }
+            FAVORITE -> swipeRefresh.isEnabled = false
```
Not clear why it's here. Could you elaborate, please?

Commit:
- edited the check on null
- remove assent from UseCase
- adeed function process() in Result.kt
- edited loadNewMovies() and loadNewMovieDitails() functions

Commit:
- rename assent for safeLet
